### PR TITLE
First steps toward implementing change suggestions

### LINKED ISSUE #1: A Scanner for Flair
Due: Fri, Sep 23

http://cs.uni.edu/~wallingf/teaching/cs4550/compiler/project01.html

### PR DESCRIPTION
**Pull request by rjpepmeyer**
Closes #1 

This does:
- Introduce an enum value, tokenType, that represents token types
- Change getType functions to return tokenType
- Provide a way to take a tokenType and return a string (called toString())
- Implement structures for tokens (token) and token lists (tokenList)
- Add the relevant values to a token, and read them from that token
- Separate the utility functions into a separate file
- Implement the linked list of tokenList
- Change main() to a function that takes as an argument a source file and return a linked list of tokenList
- Add a main file which calls the function from scanner
- Handles one-token files correctly
- Rejects numbers larger than 2^32-1
- Handles punctuation character `.` correctly
1. Compile with `g++ -std=c++14 compiler.cpp -o flairs`
2. Run with `./flairs name-of-input-file.flr`

### CODE REVIEW

File: scanner-utilities.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    if (input.length() == 1) {
+      return number;
+    }
+    else {
+      return invalidnum;
+    }
+  }
+  else if (isdigit(input.at(0))) {
+    int i = 0;
+    if (input.length() < 10 || (input.length() == 10 &&
```
Comment (dmnemec) [inline scanner-utilities.cpp]: What is 10 referencing here again? Would this be more appropriate as a function variable or something else?

File: scanner-utilities.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    else {
+      return invalid;
+    }
+  }
+  else {
+    if (input.length() > 255) {
+      return invalidnum;
+    }
+    else {
+      if (!isalpha(input.at(0))) {
```
Comment (dmnemec) [inline scanner-utilities.cpp]: Can we use input.front() instead or input.at(0)?

File: scanner-utilities.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  }
+  else if (input.at(0) == '0') {
+    if (input.length() == 1) {
+      return number;
+    }
+    else {
+      return invalidnum;
+    }
+  }
+  else if (isdigit(input.at(0))) {
```
Comment (dmnemec) [inline scanner-utilities.cpp]: input.first() instead of input.at(0) again?

File: scanner-utilities.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  else if (input == "and" || input == "or" || input == "not") {
+    return equals;
+  }
+  else if (input == "if" || input == "then" || input == "else"
+  || input == "print" || input == "program" ||
+  input == "function" || input == "return" || input == "begin"
+  || input == "end" || input == "end") {
+    return keyword;
+  }
+  else if (input.at(0) == '0') {
```
Comment (dmnemec) [inline scanner-utilities.cpp]: input.first() instead of input.at(0) again?

File: scanner-utilities.cpp [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+      if (i == input.length()) {
+        return number;
+      }
+    }
+    else {
+      return invalid;
+    }
+  }
+  else {
+    if (input.length() > 255) {
```
Comment (dmnemec) [inline scanner-utilities.cpp]: Move 255 to a "maxNumLength" or something similar?

**Review by dmnemec — COMMENTED** [review]
Looks good so far. I haven't run the code, but as far as I saw the logic was good and clean. A few minor suggested changes, but otherwise approved.

File: scanner-utilities.cpp (hunk 1 above)
Comment (rjpepmeyer) [reply scanner-utilities.cpp]: 10 is the length of 2^32-1 as a string. This is not clear at all and I'm going to move this logic into a separate function to make it more clear.

**Review by dmnemec — APPROVED** [review]
Nice changes. logic's easier to follow with number validation it's own function.

**Pull request merged**